### PR TITLE
GraphOS authorization directives: policy directive

### PR DESCRIPTION
**Pull request by Geal**
replaces #3583
~Blocked on #3647~

This introduces the `@policy` directive, defined as follows:

```graphql
scalar federation__Policy
directive @policy(policies: [[federation__Policy!]!]!) on
    OBJECT
  | FIELD_DEFINITION
  | INTERFACE
  | SCALAR
  | ENUM
```

It evaluates a list of policies, represented as arbitrary strings, that are stored in the context as a map String -> Option<bool>. A coprocessor or rhai script at the supergraph stage will look at this list of policies, and indicate in the map which were successful. The router will then accept fields for which all the required policies are successful.
The nested array in the `policies` argument works in the same way as `@requiresScopes`: the inner array is an AND between policies, all of them have to be successful at the same time, while the outer array is an OR between the inner arrays, at least one of these inner arrays must be successful.

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

**Review by bnjjj — APPROVED** [review]

File: docs/source/configuration/authorization.mdx [hunk 1]
```diff
@@ -395,6 +396,165 @@ The response retains the initial request's shape but returns `null` for unauthor
...
+  };
+  service.map_request(request_callback);
+}
+```
+
+The script uses the [`eval` function](https://rhai.rs/book/ref/eval.html) to evaluate each policy and store the result in the `apollo_authorization::policies::required` map.
+
+</ExpansionPanel>
+
+##### Usage with a coprocessor
```
Comment (garypen) [inline docs/source/configuration/authorization.mdx:475]: Same comment about RouterRequest stage.

File: docs/source/configuration/authorization.mdx [hunk 2]
```diff
@@ -395,6 +396,165 @@ The response retains the initial request's shape but returns `null` for unauthor
...
+@policy(policies: ["claims[`roles`].contains(`support`)"])
+```
+
+The preceding example includes one policy that requires all roles from the claims object to include the string `"support"`.
+
+The Apollo Router extracts the list of policies relevant to a request from the schema. It then stores them in the request's context in `apollo_authorization::policies::required` as a map `policy -> null|true|false`. This process happens at the [`RouterService` level](../customizations/overview#the-request-lifecycle).
+
+You need to provide a Rhai script or a coprocessor at the `SupergraphService` level that evaluates the map, setting the value to `true` if the policy is validated and `false` if not. After that, the router filters the requests' types and fields to only those where the policy is `true`.
+
+> If the `policies` array contains multiple elements, _only one must be successful for the policy to be `true`_.
```
[the remark this answers is not in the document]
Comment (Geal) [reply docs/source/configuration/authorization.mdx:415]: the assumption was that AND would be inside the policy implementation, but maybe it should just align here :thinking:

File: docs/source/configuration/authorization.mdx [hunk 3]
```diff
@@ -395,6 +396,165 @@ The response retains the initial request's shape but returns `null` for unauthor
...
+To use the `@policy` directive in a subgraph, you can [import it from the `@link` directive](/federation/federated-types/federated-directives/#importing-directives) like so:
+
+```graphql
+extend schema
+  @link(
+    url: "https://specs.apollo.dev/federation/v2.5",
+    import: [..., "@policy"])
+```
+
+Using the `@policy` directive requires a [Supergraph plugin](../customizations/overview) to evaluate the authorization policies. You can do this with a [Rhai script](../customizations/rhai/), [coprocessor](../customizations/coprocessor), or [native plugin](../customizations/native). Refer to the following Rhai script and coprocessor examples for more information.
```
[the remark this answers is not in the document]
Comment (Geal) [reply docs/source/configuration/authorization.mdx:430]: it uses data extracted by the query analysis layer, which runs inside the router service, so this task has to happen in a supergraph coprocessor

File: docs/source/configuration/authorization.mdx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Geal) [reply docs/source/configuration/authorization.mdx]: done in https://github.com/apollographql/router/pull/3751/commits/55f722e3fc73be31f809710423dab12da07b701a

**Pull request merged**